.github/type_check.py

```diff
@@ -1,15 +1,60 @@
+import json
 from pathlib import Path
 import subprocess
+from typing import Any, Dict, List, Union
 
-config = Path(__file__).parent / "pyright-config.json"
+source_dir = Path(__file__).parent
+config = source_dir / "pyright-config.json"
 
-command = ("pyright", "-p", str(config))
-print(" ".join(command))
 
-try:
-    result = subprocess.run(command)
-except FileNotFoundError as e:
-    print(f"{e} - Is pyright installed?")
-    exit(1)
+def pyright() -> int:
+    """ returns process return code """
+    command = ("pyright", "-p", str(config))
+    print(" ".join(command))
 
-exit(result.returncode)
+    try:
+        pyright_result = subprocess.run(command)
+    except FileNotFoundError as e:
+        print(f"{e} - Is pyright installed?")
+        exit(1)
+    return pyright_result.returncode
+
+
+def mypy() -> int:
+    """ returns process return code """
+    with open(config) as config_file:
+        config_data: Union[Dict[str, Any], Any] = json.load(config_file)
```

Comment:
Shouldn't mypy use a separate list of files, so as the maintainer of a code snippet can decide which one they think should be applied for their code (assuming they are not 1:1 compatible)?

Reply:
We can't really isolate the checks to a specific list of files.
The list of files we give (to either mypy or pyright) is not a list of files to check, It's a list of files to report errors in. Both type checkers follow imports to get type information from other files. So the type information in any file is going to affect both checks.
If someone thinks they're only using mypy, because the file they're changing is only on the mypy list, that change will still affect the pyright check, even though the file is not on the pyright list.

The main focus of this is core, and worlds are opt-in.
If this PR is accepted (I don't consider it decided yet whether we want this), If a world maintainer wants to only use one, then they shouldn't opt in to this system, and should just check with their own workflow.

Reply:
You shouldn't think of this as a linter, checking stylistic things.
You should think of it as a static type checker, just like in a C++ or Rust compiler.

I think it wouldn't make sense to say:
"For this C++ file `x.cpp` I only care whether clang's type checker passes, but for `y.cpp` I only care whether gcc's type checker passes."
We would care about everything passing, in one compiler, or the other compiler, or we want to support both compilers.

I think it would be good to support multiple static type checkers, just like a C++ project that wants to support multiple compilers.

Reply:
That argument makes sense, yeah.

That being said, I honestly don't know how we (or who) decide(s) if we want to merge this or not.